Yoast/wordpress-seo
# Add black friday checklist promotion in product editor

## Context


* We want to show a promotion notice in the WooCommerce product editor.

## Summary


This PR can be summarized in the following changelog entry:

* Adds a reusable, time-bounded notification to advertise promotions.
* Shows a promotion notice in the product editor.

## Relevant technical choices:

* A `TimeConstrainedNotice` reusable component has been introduced: it relies on the `Promotions_Manager` to decide whether the notice should be shown.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:
#### Preliminary steps
* Test this PR on a website which has an unauthorizer Yoast SEO Premium
* Activate WooCommerce
* Deactivate Yoast SEO Premium
* Edit `src/promotions/domain/black-friday-checklist-promotion.php`
  * change `\gmmktime( 11, 00, 00, 9, 19, 2023 )` to `\gmmktime( 11, 00, 00, 9, 1, 2023 )`

#### Check the promotion notice is correctly shown
* Go to `Products` -> `Add new`
* In the metabox, check that you see the following notice:
  <img width="630" alt="Screenshot 2023-09-05 at 00 03 51" src="https://github.com/Yoast/wordpress-seo/assets/68744851/0e9e4996-85e3-44d6-a5d9-0b5d424d0a8f">
* Edit a post other than a `Product`
  * verify the notice is not shown

#### Test the notice is not shown when a warning is present in the metabox
* Now activate  Yoast SEO Premium
* Go to `Products` -> `Add new`
* In the metabox, check that you don't see the previous notice, and see the following warning instead:
  <img width="628" alt="Screenshot 2023-09-05 at 00 33 29" src="https://github.com/Yoast/wordpress-seo/assets/68744851/5f6146cd-5414-40e7-b305-d94dfa303d9f">

#### Test the notice is dismissed correctly
* Deactivate  Yoast SEO Premium
* Go to `Products` -> `Add new`
* Dismiss the notice and refresh the page
  * verify the notice doesn't show up again 

#### Test the notice is not shown outside its time frame
* With a MySQL client edit the `wp_usermeta` table
  * search  for the tuple with `user_id` = your user id and `meta_key` = `_yoast_alerts_dismissed`
    * set `meta_value` to `a:0:{}` 
* Go to `Products` -> `Add new` and verify the notice shows correctly
* Edit `src/promotions/domain/black-friday-checklist-promotion.php`
  * change `\gmmktime( 11, 00, 00, 11, 25, 2023 )` to `\gmmktime( 11, 00, 00, 9, 3, 2023 )`
* Go to `Products` -> `Add new` and verify the notice is not shown

#### Relevant test scenarios


### Test instructions for QA when the code is in the RC



## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

* N/A

## UI changes


## Other environments


## Documentation


## Quality assurance


## Innovation


Fixes https://github.com/Yoast/ux/issues/67
